akka-http/src/main/scala/akka/http/scaladsl/server/directives/WebSocketDirectives.scala

```diff
@@ -76,4 +80,43 @@ trait WebSocketDirectives {
       else
         reject(UnsupportedWebSocketSubprotocolRejection(subprotocol.get)) // None.forall == true
     }
+
+  /**
+   * Checks that the WebSocket comes from the same origin. Extracts the [[Origin]] header and
+   * verify that the allowed range contains obtained values. In case of absent of [[Origin]] header
+   * rejects with [[MissingHeaderRejection]] and [[StatusCodes.Forbidden]]. In case if some of the origin
+   * values not in the allowed list rejects with a [[MalformedHeaderRejection]] and [[StatusCodes.Forbidden]].
+   *
+   * See https://tools.ietf.org/html/rfc6455#section-1.3 and
+   * http://blog.dewhurstsecurity.com/2013/08/30/security-testing-html5-websockets.html
+   *
+   * @group websocket
+   */
+  def checkSameOrigin(allowed: HttpOriginRange): Directive0 = {
```

Comment:
Impl pretty good, however it's not a websocket specific directive, I'd rather put it under HeaderDirectives as we operate (require) as specific Header (Origin).